alonzo/formal-spec/properties.tex

```diff
@@ -197,45 +197,210 @@ \section{Formal Properties}
+\end{itemize}
+
+The fixed argument values provided by the ledger, together with the assumption that there
+are no additional data the interpreter can inspect, allows us to say that
+script evaluation is entirely deterministic. More precisely, we say that
+
+\begin{lemma}
+  For all environments $e, e'$, transactions $tx, tx'$ and states $s, s', u, u'$ such that
+  \begin{equation*}
+    e\vdash s\trans{ledger}{tx}s', \\
+    e'\vdash u\trans{ledger}{tx'}u', \\
+    \txbody{tx} = \txbody{tx'}, \\
+    (\fun{languages}~(\fun{txwits}~{tx}))~\restrictdom~\fun{costmdls}~(\field_{pp}(s))~=~(\fun{languages}~(\fun{txwits}~{tx'}))~\restrictdom\fun{costmdls}~(\field_{pp}(s')), \\
+  \end{equation*}
+  The following holds :
+
+  \[\fun{evalScripts}~{tx}~ \fun{collectTwoPhaseScriptInputs} ~\field_{pp}(s)~\var{tx}~ \Utxo(s)\]
+  \[ \Leftrightarrow\]
+  \[\fun{evalScripts}~{tx'}~ \fun{collectTwoPhaseScriptInputs} ~\field_{pp}(u)~\var{tx'}~ \Utxo(u)\]
+
+\end{lemma}
+\begin{proof}
+  Let us consider the use of arguments of $\fun{evalScripts}$ (see Figure \ref{fig:functions:script2}).
+  \begin{itemize}
+    \item $\var{tx}, \var{tx'}$ : this argument is only inspected in the case that the script $sc$ (the first element
+    in the pair at the head of the list
+    $((\var{sc}, \var{d}, \var{eu}, \var{cm});\Gamma)$) is a native script. Since all native scripts
+    are checked in phase one of validation (see \ref{fig:rules:utxow-alonzo}) by calling $\fun{validateScript}$
+    on all scripts attached to the transaction. For this to apply to $sc$, we must also show
+    that $sc$ is a script attached to the transaction (see the second argument explanation).
+    Note also that the $tx$ argument passed to $\fun{evalScripts}$ at the use site (in the UTXOS transition)
+    is the same, unmodified $tx$ as is the signal for the LEDGER transition. We verify this by inspecting
+    the sequence of transitions through which it is propagated (UTXOS, UTXO, UTXOW, and LEDGER), and their signals.
+
+    This allows us to conclude that the first argument to $\fun{evalScripts}$
+    \begin{itemize}
+      \item has no impact on the outcome of the evaluation of the non-native scripts
+      as part of running $\fun{evalScripts}$, and
+      \item the outcome of native script evaluation by $\fun{evalScripts}$ is the
+      same as the outcome of evaluating them in phase-1, since
+      it calls the same evaluation function on the same arguments in both cases
+      \[\fun{evalTimelock}~\var{tx}~\var{sc}\]
+      Note here that we are considering only phase-1 valid transactios here (as we started
+      with the assumptions of having valid ledger transitions), therefore
+      phase-1 evaluation of native scripts necessarily succeeded for both $tx$ and $tx'$
+      in order to proceed to phase-2, wherein $\fun{evalScripts}$ is run
+    \end{itemize}
+
+    \item The second argument is a list. We will show that it must be the same list for the $e, s, tx, s'$
```

Comment:
This proof seems to be mostly a repetition of the properties already discussed in the main text. I think this should be more formal instead.  Also, I don't think it's necessary to remind the reader about what a function does here. They probably need to refer to its definition anyway if they don't remember.